src/Nethermind/Nethermind.Core/Collections/SpanDictionary.cs

```diff
@@ -0,0 +1,1639 @@
+                    catch (InvalidCastException)
+                    {
+                        throw new ArgumentException($"The value \"{value}\" is not of type \"{typeof(TValue)}\" and cannot be used in this generic collection.", nameof(value));
+                    }
+                }
+                catch (InvalidCastException)
+                {
+                    throw new ArgumentException($"The value \"{key}\" is not of type \"{typeof(TKey[])}\" and cannot be used in this generic collection.", nameof(key));
+                }
+            }
+        }
+
+        private static bool IsCompatibleKey(object key)
+        {
+            ArgumentNullException.ThrowIfNull(key);
+            return key is TKey[];
+        }
+
+        void IDictionary.Add(object key, object? value)
+        {
+            ArgumentNullException.ThrowIfNull(key);
+            ThrowHelper.IfNullAndNullsAreIllegalThenThrow<TValue>(value, nameof(value));
+
+            try
+            {
+                TKey[] tempKey = (TKey[])key;
+
+                try
+                {
+                    Add(tempKey, (TValue)value!);
+                }
+                catch (InvalidCastException)
+                {
+                    throw new ArgumentException($"The value \"{value}\" is not of type \"{typeof(TValue)}\" and cannot be used in this generic collection.", nameof(value));
+                }
+            }
+            catch (InvalidCastException)
+            {
+                throw new ArgumentException($"The value \"{key}\" is not of type \"{typeof(TKey[])}\" and cannot be used in this generic collection.", nameof(key));
+            }
+        }
+
+        bool IDictionary.Contains(object key)
+        {
+            if (IsCompatibleKey(key))
+            {
+                return ContainsKey((TKey[])key);
+            }
+
+            return false;
+        }
+
+        IDictionaryEnumerator IDictionary.GetEnumerator() => new Enumerator(this, Enumerator.DictEntry);
+
+        void IDictionary.Remove(object key)
+        {
+            if (IsCompatibleKey(key))
+            {
+                Remove((TKey[])key);
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private ref int GetBucket(uint hashCode)
+        {
+            int[] buckets = _buckets!;
+            return ref buckets[HashHelpers.FastMod(hashCode, (uint)buckets.Length, _fastModMultiplier)];
+        }
+
+        private struct Entry
+        {
+            public uint hashCode;
+            /// <summary>
+            /// 0-based index of next entry in chain: -1 means end of chain
+            /// also encodes whether this entry _itself_ is part of the free list by changing sign and subtracting 3,
+            /// so -2 means end of free list, -3 means index 0 but on free list, -4 means index 1 but on free list, etc.
+            /// </summary>
+            public int next;
+            public TKey[] key;     // Key of entry
```

Comment:
This is the place where the key needs to be allocated. This means that for non existing keys, we'll get the allocations here. Depending on the case, how likely it's going to happen this could be ok or not. This is the reason for bringing an idea of having a struct that represents the key if we know the upper boundary on the lenght of the key and use a regular `ConcurrentDictionary` for it. I assume that it will be Keccaks only, but maybe there are other cases.